### ISSUE TITLE
Uninformative printing for :uconstr:(id (ltac:(exact tt))) and ltac:(auto with *)

### ISSUE THREAD

**Issue by jfehrle** [coq/coq#14605]
While discussing how to display variable values in the debugger, @JasonGross mentioned that some constructs are currently not printed in a useful form.  Here are two examples:

```
Goal True.
Ltac bug1 := let x := uconstr:(id (ltac:(exact tt))) in idtac x.
bug1.    (* prints "(id _)" *)

Ltac bug2 := let t := ltac:(auto with *) in idtac t.
bug2.    (* prints "<tactic closure>" *)
```

@herbelin, it would be great if you can make these show more useful values.

Note that the new debugger code calls `Pptactic.pr_value Constrexpr.LevelSome <value>`, which gives the same result that `idtac` prints in `bug1`, but gives "\<genarg:tacvalue\>" for `bug2`, as shown in the "Variables" panel of this screen shot:

![image](https://user-images.githubusercontent.com/1253341/124507848-8662ac80-dd83-11eb-8ca5-55baf4f3f69a.png)

**Comment by herbelin** [comment]
Incidentally, another instance of the first example above (since `$id` use the quotation mechanism):
```
From Ltac2 Require Import Ltac2.
Ltac2 f x := constr:(S $x).
Print Ltac2 f.
(*
f : constr -> constr
f := fun x => constr:(S _)
*)
```
So that all constr quotations are parsed and printed consistently, we may propose a new mechanism, such that we could do:
```
CONSTR EXTEND
  INTERPRETED BY { interp_ltac2_quotation }                                                                                          
  GLOBALIZED BY { intern_ltac2_quotation }                                                                                             
  SUBSTITUTED BY { subst_ltac2_quotation }                                                                                           
[ "$" lident(x) ] -> { Genarg.in_gen (Genarg.rawwit wit_ltac2_quotation) id }
END
```
and
```
CONSTR EXTEND
  INTERPRETED BY { interp_ltac2_quotation }                                                                                          
  GLOBALIZED BY { intern_ltac2_quotation }                                                                                             
  SUBSTITUTED BY { subst_ltac2_quotation }                                                                                           
[ "ltac" ":" "(" ltac_expr(tac) ")" ] -> { Genarg.in_gen (Genarg.rawwit wit_tactic) tac }
END
```
or another mechanism which supports only quotations of the form `foo:(some-text)`?

@ppedrot: Do you have an opinion or suggestion to make on this question? Or are you ok if ever I proceed one way or another? Thank in advance for your help.

**Comment by herbelin** [comment]
A risk with the second example:
```
Ltac bug2 := let t := ltac:(auto with *) in idtac t.
bug2.    (* prints "<tactic closure>" *)
```
is that the code to print may quickly come very long. 

I made `https://github.com/herbelin/github-coq/tree/master+printing-ltac-value-full` as an experiment. Here is an example of what it gives:
```
Set Ltac Debug. 
Goal True.
Ltac f tac x := let t := tac x in idtac t.
f ltac:(fun x => assert x) True. 
```
eventually shows:
```
Debug:
Level 1: evaluation returns
idtac t
where
x := True
tac := fun x => assert x
t := assert x
     where
     x := True
```
(plus a repetition)

**Comment by jfehrle** [comment]
> the code to print may quickly come very long

Indeed.  I've already updated the Variables panel to use a tree control, in which every variable starts with a single line, then you to click to display any multiline values.  It looks like this now:

![image](https://user-images.githubusercontent.com/1253341/125543219-5b625af4-c5b6-4cb9-8d67-4b6a47d6f2f9.png)

CoqIDE responds almost instantaneously even when some variable values are several pages long.  That's probably helped by the changes I make in [this commit](https://github.com/coq/coq/pull/14644/commits/6107685f4b3d9c3dbd52639936fb0724e5c2b133), which display feedback immediately rather than queueing it and printing it later printed from a 300 msec timer.

**Comment by jfehrle** [comment]
> I made master+printing-ltac-value-full as an experiment. Here is an example of what it gives:

That looks very good.  `x` and `tac` exactly right.  For `t`, is it possible to print it as `let x := True in assert x`?  @JasonGross suggested that it would be most helpful if the printed values were also valid Coq inputs, which seems quite sensible to me.

FWIW, when I clicked on the link to your branch, I couldn't see the code.

**Comment by jfehrle** [comment]
Perhaps `tac` should be printed as `ltac:(fun x => assert x)`?  @JasonGross?

**Comment by JasonGross** [comment]
> FWIW, when I clicked on the link to your branch, I couldn't see the code.

You can get a code comparison by going to https://github.com/coq/coq/compare/master...herbelin:master+printing-ltac-value-full?expand=1 and scrolling down

> Perhaps `tac` should be printed as `ltac:(fun x => assert x)`?

No, I don't think so.  The default tactic scope renders `fun x => assert x` as valid (`Ltac foo := let tac := fun x => assert x in tac.` is fine).  It may be worth somehow printing what sort of object each variable is, or having a printing option that wraps things in `ltac:(...)`, `constr:(...)`, `uconstr:(...)`, etc?  Usually I think that having `ltac:(...)` or `constr:(...)` around everything will just be extra line noise, though I can imagine rare cases where not having this information results in a lot of head-scratching.

**Comment by jfehrle** [comment]
> It may be worth somehow printing what sort of object each variable is, or having a printing option that wraps things in ltac:(...), constr:(...), uconstr:(...), etc? 

Would it work to append strings such as `: Prop` to the value?  Not sure if that's valid syntax or not.  And perhaps omit the sort for the most common case for less visual clutter?  That would be `Type`?  There could be a separate routine the debugger could call to fetch the sort of an expression--or it could be added to the printer behavior.

At the moment, there's no support for changing option values while you're in the debugger, nor can you enter "Set" commands in the Message window.  We could have GUI menu items for the former.  The latter could be done by interfacing to the main parser and filtering to accept only the appropriate (small) subset.

**Comment by JasonGross** [comment]
> Would it work to append strings such as `: Prop` to the value? Not sure if that's valid syntax or not. And perhaps omit the sort for the most common case for less visual clutter? That would be `Type`?

No, this is not the sort of type I mean.  I mean basically the thing that gets printed with `pp_type` in
https://github.com/coq/coq/blob/f684f34eed7889cbced249630e4ffe9306b1401b/plugins/ltac/tacinterp.ml#L236-L250
Another way of pointing at roughly the same thing is that there's a bunch of `FOO:(...)` constructs for different values of `FOO` (`ltac`, `uconstr`, `constr`, etc), and I'm talking about whichever one needs to be used to parse/interpret the object.

**Comment by herbelin** [comment]
>  For `t`, is it possible to print it as `let x := True in assert x`? 

Yes, that would be possible. Unclear though whether it is easy to determine when to insert an `ltac:(...)`, `constr:(...)` etc.

**Comment by jfehrle** [comment]
@herbelin What should I do with `master+printing-ltac-value-full`?  Copy those changes and see how well it works in the debugger?  If it looks good, could we then submit it as a PR?

**Comment by jfehrle** [comment]
@herbelin Now that the debugger is in 8.15, it's a good time for someone to address this bug.  Perhaps you would take that on?  IIUC you are the expert on printing.  I looked at the printing code a couple months ago but wasn't able to figure out much of how it works--it seemed like it would take me quite a long time to do so.

You earlier mentioned your experiment `master+printing-ltac-value-full` that worked for the cases I tried.

WDYT?

**Comment by herbelin** [comment]
The best I could do in the short term is a PR from the branch if it can help you.

**Comment by jfehrle** [comment]
Great, that would be a useful step.  Then we can update this PR or create a new PR describing any remaining issues (which perhaps could be addressed for 8.16).
